### DISCUSSION TITLE
Wallet API redesign

### DISCUSSION THREAD

**Discussion by genaris** [openwallet-foundation/credo-ts#1486] [Architecture & Design]
Agent Wallet API is currently based on the means that Indy SDK provided, and makes it a bit complicated to handle different scenarios, especially when using new wallet types that do not follow the same lifecycle and kind of parameters than it.

Some ideas were given at June 15th 2023 AFJ WG call to split out the responsibilities between what's in the "Wallet module" (Agent Core module), the "Wallet provider module" (module that provides the actual implementation, such as AskarModule or IndySdkModule) and the Wallet object itself.

Link to the slides shown at today's AFJ WG call: https://docs.google.com/presentation/d/1KM-jAIgMQrJL207YSY-Lvdh1FISI0gBAp5bW4WGHaOQ/edit?usp=sharing

Some related issues:
- https://github.com/hyperledger/aries-framework-javascript/issues/1236
- https://github.com/hyperledger/aries-framework-javascript/issues/1302

**Comment by TimoGlastra** [comment]
I'm working on adding supporting for using the same store but with different profiles for Askar in multi-tenancy. I'm going to implement it for 0.4.x without breaking changes for now, but this introduces some 'hacks'. 

I think the following change would be really useful to make it better suitable for this (from #1236):
> maybe WalletConfig can be left as generic as possible and leave each Wallet kind to define their own interfaces to identify and configure the wallets

Not having to deal with a generic wallet interface, but being able to have a custom interface makes sense, and also would accomodate better to e.g. an in memory wallet that does not require an id and key, or key derivation method.

In addition the following would also be really useful:
> split out the responsibilities between what's in the "Wallet module" (Agent Core module), the "Wallet provider module" (module that provides the actual implementation, such as AskarModule or IndySdkModule) and the Wallet object itself.

Extracting the wallet management into a separate wallet provider class/module will help a lot, as the wallet implementation part for Askar is the same, but there's difference in whether we want to use profiles for tenants or not for example.

**Comment by TimoGlastra** [comment]
Done :)
